providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,300 @@
+                                HYBRID_PKEY *authpriv,
+                                HYBRID_PKEY *verify,
+                                const OSSL_PARAM params[],
+                                int (*initf)(EVP_PKEY_CTX *ctx,
+                                             EVP_PKEY *authpriv,
+                                             const OSSL_PARAM params[]))
+{
+    unsigned int i;
+
+    if (!ossl_prov_is_running() || !ossl_hybrid_kem_set_ctx_params(ctx, params))
+        return 0;
+    for (i = 0; i < ctx->num_algs; i++)
+        if (!(*initf)(ctx->ctxs[i], authpriv->keys[i], params))
+            return 0;
+    return 1;
+}
+
+int ossl_hybrid_kem_auth_encapsulate_init(void *vctx, void *vrecip,
+                                          void *vauthpriv,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *authpriv = (HYBRID_PKEY *)vauthpriv;
+
+    return hybrid_kem_auth_init(ctx, authpriv, NULL, params,
+                                &EVP_PKEY_auth_encapsulate_init);
+}
+
+int ossl_hybrid_kem_auth_decapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpriv,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *authpriv = (HYBRID_PKEY *)vauthpriv;
+
+    return hybrid_kem_auth_init(ctx, authpriv, key, params,
+                                &EVP_PKEY_auth_decapsulate_init);
+}
+
+int ossl_hybrid_kem_get_ctx_params(void *vctx, OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_get_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+const OSSL_PARAM *ossl_hybrid_kem_settable_ctx_params(void *vctx, void *provctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_settable_ctx_params(ctx, provctx);
+}
+
+const OSSL_PARAM *ossl_hybrid_kem_gettable_ctx_params(void *vctx, void *provctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_gettable_ctx_params(ctx, provctx);
+}
+
+int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t total_len = ossl_hybrid_get_key_length_from_ctx(ctx);
+    size_t a, b, encsum = 0, secretsum = 0;
+    size_t elen[MAX_HYBRID_ALGS], slen[MAX_HYBRID_ALGS];
+    unsigned int i;
+
+    if (total_len == 0)
+        return 0;
+    /* Calculate lengths */
+    for (i = 0; i < ctx->num_algs; i++) {
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], NULL, i + elen,
```

Comment:
elen and slen are strange names.. Even stranger is the order you do i + elen which makes it not look like an array offset

Reply:
enclen and secretlen were taken...